### ISSUE TITLE
Move property literals out of Plotly.restyle

### ISSUE THREAD

**Issue by monfera** [plotly/plotly.js#648]
See items 1, 2, 3 in https://github.com/plotly/plotly.js/pull/617#discussion_r67130816. 
Preceding comment from @etpinard: https://github.com/plotly/plotly.js/pull/617#discussion_r67041656

**Comment by etpinard** [comment]
I think (at least I like to think) that there's a way to achieve this using our attribute declaration. 

Some questions that we need to answer:

- Do we need all restyle `docalc`, `doplot`, ... [distinctions](https://github.com/plotly/plotly.js/blob/9ee4525189b5377cebc35f691e692f7865115ff2/src/plot_api/plot_api.js#L1638-L1643) ? Can you merge some of them?

- Most `docalc` updates relate to our `data_array` and our `arrayOk` attributes. Meaning that we could simply use the attribute declaration to figure our what update path to take. Are there any exceptions to that rule?

- _anything else ?_

**Comment by etpinard** [comment]
From @rreusser 's https://github.com/plotly/plotly.js/issues/1578 

> Carpet is going to add a huge number of attrs to recalcAttrs. (up to 287 total or so!?) The array of attributes is recalculated on every plot and it looks them up with indexOf. It should place them in a hash (e.g. {mode: true, visible: true, ...}) and should be defined in a plot-api-global context since it never changes.

> @alexcjohnson notes that a larger refactoring is desired, but this might be a start since, at least once carpet is merged, it shouldn't take more than five minutes and may very well be at the point of having a non-negligible impact on performance.

**Comment by etpinard** [comment]
A pretty valuable first step was made in https://github.com/plotly/plotly.js/pull/1653/files/413e5459960c4b4c2c4c6212919e3a5c3183f981#diff-2941ab69a12080c0633ff4ac8ea3aa83

**Comment by etpinard** [comment]
https://github.com/plotly/plotly.js/pull/1999 will probably close this.

**Comment by etpinard** [comment]
Superseded by https://github.com/plotly/plotly.js/issues/2025
